kube-proxy: internal config: refactor BindAddress

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
This PR replace BindAddress with NodeIPOverride for internal configuration of kube-proxy adhering to the v1alpha2 version specifications as detailed in https://kep.k8s.io/784.


split out from https://github.com/kubernetes/kubernetes/pull/121830/commits
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman commented:
/sig network
/cc @danwinship

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -227,6 +232,13 @@ func (o *Options) Complete(fs *pflag.FlagSet) error {
+// processV1Alpha1Flags processes v1alpha1 flags which can't be directly mapped to internal config.
+func (o *Options) processV1Alpha1Flags(fs *pflag.FlagSet) {
+	if fs.Changed("bind-address") {
+		o.config.NodeIPOverride = []string{o.bindAddress}
```
So I'm not sure that `--bind-address X` and `nodeIPOveride: [X]` should mean the same thing...

In particular, `--bind-address X` means "I want X to be the primary IP, but maybe there's also another IP". But with `nodeIPOverride`, it seems weird to have a single IP imply "but maybe there's also another IP". The whole point of it being an array is so you can specify both, if you want both. So fundamentally, `nodeIPOverride: [X]` feels to me like it should means "X is the only IP". (And, for comparison, that's how it works with `kubelet --node-ip`; we autodetect 1 or 2 IPs if you specify nothing, but if you say `kubelet --node-ip 192.168.1.2`, that means _just_ `192.168.1.2`; you have to specify both IPs explicitly if you want non-autodetected dual-stack.)

But in turn, that implies that you can't just convert `--bind-address` to `nodeIPOverride`... but also, we don't want there to be functionality available via the old API that's not available via the new API...

So, OK, two possibilities:
1. You can say something like `nodeIPOverride: ["192.168.1.2", "::"]` to mean "the node IPs are 192.168.1.2 and whatever IPv6 IP you want"
2. We add an `ipFamilyPolicy` field like in `Service`, which defaults to `PreferDualStack` if you don't specify `nodeIPOverride`, `SingleStack` if you specify a single `nodeIPOverride` value, and `RequireDualStack` if you specify 2 `nodeIPOverride` values. And then `--bind-address 192.168.1.2` would be translated to `nodeIPOverride: ["192.168.1.2"], ipFamilyPolicy: PreferDualStack`

I think I like the latter better.

(But I think we need to think about this for longer than 8 hours, so this isn't making it into 1.31.)

aroradaman replied:
After https://github.com/kubernetes/enhancements/pull/5064 I'm not really sure how to consume NodeIPOverride. It will work well for nodePortAddresses but currently Healthz and Metrics server address default values depend on BindAddress. 

https://github.com/kubernetes/kubernetes/blob/06331ac573a182bdace8f238ca94a1f476958b21/pkg/proxy/apis/config/v1alpha1/defaults.go#L142-L150

Given now healthz and metrics server will be single-stack, defaulting the values will be difficult as nodeIPOverride can be a dual stack ip address with ipFamilyPolicy PreferDualStack.

Alternatively, we can make healthz and metrics server dual stack, not exactly dual-stack CIDRs which broke the CI but dual-stack IPs which can be defaulted using NodeIPOverride?

aojea replied:
this is the same discussion as before, these are control plane endpoints, and dual stack is usually referring to the dataplane, as in I can have Services with IPv4 and IPv6 addresses, but people don't deploy dual stack monitoring applications, as prometheus for IPv6 and prometheus for IPv4

aroradaman replied:
Should I remove that from the design as well then?

danwinship replied:
though I guess `0.0.0.0` vs `::` doesn't really matter... so maybe we should continue to default those values at defaulting time, and just base it on `nodeIPOverride[0]` rather than `bindAddress`. (Defaulting to IPv4 if `nodeIPOverride` wasn't set.)

aroradaman committed:
kube-proxy: internal config: refactor BindAddress

Replace BindAddress with NodeIPOverride for internal configuration of
kube-proxy adhering to the v1alpha2 version specifications as detailed
in https://kep.k8s.io/784.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
cmd/kube-proxy: fix NodeIP detection doc

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/reopen
/remove-lifecycle rotten

aroradaman commented:
/assign @danwinship @aojea

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -628,7 +628,7 @@ func (s *ProxyServer) birthCry() {
 // (Note that on Linux, the node IPs are used only to determine whether a given
 // LoadBalancerSourceRanges value matches the node or not. In particular, they are *not*
-// used for NodePort handling.)
+// used for NodePort handling unless NodePortAddresses is not set to 'primary'.)
```
The "not" is incorrect. They're used if NodePortAddresses *is* set to 'primary', not if it's *not* set to 'primary'.

But it would be better to rewrite the whole comment.

"(Note that on Linux, the node IPs are used only (a) to determine whether a given LoadBalancerSourceRanges value matches the node or not, and (b) as the IP(s) to accept NodePort connections on when NodePortAddresses is set to 'primary'.)"

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -302,3 +307,23 @@ func (m *LocalMode) Type() string {
+	// IPFamilyPolicyRequireDualStack indicates that kube-proxy is preferred to work with NodeIPs
+	// of both the IP families.
+	IPFamilyPolicyRequireDualStack IPFamilyPolicy = "RequireDualStack"
+)
```
You don't need to duplicate these here; you can just reuse `v1.IPFamilyPolicy` in the KubeProxyConfiguration.

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -179,10 +179,15 @@ type KubeProxyConfiguration struct {
+	// primary IPs are. nodeIPOverride[0] will be used for defaulting healthzBindAddress
+	// and metricsBindAddress values. If nodePortAddresses is set to "primary" node-ports
+	// will be exposed on the given ips.
+	NodeIPOverride []string
```
I feel like we should be documenting what the node IP is for here (especially because we'll forget to change it later). We should just be documenting that this lets you override the node IP if kube-proxy is autodetecting the wrong IP(s).

danwinship replied:
I wrote "should" when I meant "shouldn't" but you figured out what I meant anyway :slightly_smiling_face:

danwinship commented on pkg/proxy/apis/config/types.go:
```diff
@@ -179,10 +179,15 @@ type KubeProxyConfiguration struct {
+	// ipFamilyPolicy controls the nodeIP detection logic, which further influences if LoadBalancer
+	// services should be allowed on node when SourceRanges are set and on which interfaces node
+	// ports should be exposed when nodePortAddresses is set to 'primary'.
+	IPFamilyPolicy IPFamilyPolicy
```
"ipFamilyPolicy controls whether kube-proxy runs in single-stack or dual-stack mode. The default is 'PreferDualStack' if nodeIPOverride is not specified, 'SingleStack' if nodeIPOverride contains a single IP, or 'RequireDualStack' if nodeIPOverride contains dual-stack IPs."

(But also, you don't seem to be defaulting it?)

danwinship replied:
but then it never gets defaulted for v1alpha2?

danwinship commented on cmd/kube-proxy/app/options.go:
```diff
@@ -346,6 +347,9 @@ func (o *Options) processV1Alpha1Flags(fs *pflag.FlagSet) {
 		o.config.DetectLocal.ClusterCIDRs = strings.Split(o.clusterCIDRs, ",")
 	}
+	if fs.Changed("bind-address") {
+		o.config.NodeIPOverride = []string{o.bindAddress}
```
and `o.config.IPFamilyPolicy = PreferDualStack`?

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -631,10 +631,10 @@ func (s *ProxyServer) birthCry() {
-//  1. if bindAddress is not 0.0.0.0 or ::, then it is used as the primary IP.
+//  1. if nodeIPOverride[0] is not 0.0.0.0 or ::, then ip family of nodeIPOverride[0] is considered as primary IP family.
 //  2. if rawNodeIPs is not empty, then its address(es) is/are used
 //  3. otherwise the node IPs are 127.0.0.1 and ::1
```
This is incorrect at this point. (In particular, it never says that it actually uses the IPs in nodeIPOverride as the node IPs.)

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -200,7 +200,7 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
 	rawNodeIPs := getNodeIPs(ctx, s.Client, s.Hostname)
-	s.PrimaryIPFamily, s.NodeIPs = detectNodeIPs(ctx, rawNodeIPs, config.BindAddress)
+	s.PrimaryIPFamily, s.NodeIPs = detectNodeIPs(ctx, rawNodeIPs, config.NodeIPOverride, config.IPFamilyPolicy)
 
```
at some point in `newProxyServer` you need to check if ipFamilyPolicy is RequireDualStack but !dualStackSupported.

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -657,15 +657,33 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
 		}
-		nodeIPs[primaryFamily] = bindIP
+		delete(nodeIPs, secondaryFamily)
 	}
```
This is pretty confusing, and it's hard to be sure we got all the cases right...

I think it would be clearer if you figured out the primary IP family first, and then filled in nodeIPs separately after that.

danwinship commented on cmd/kube-proxy/app/server.go:
```diff
@@ -657,15 +661,25 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
+		family := v1.IPv4Protocol
+		ip := netutils.ParseIPSloppy(nodeIPOverride[i])
+		if ip != nil && !ip.IsUnspecified() {
+			if !netutils.IsIPv4(ip) {
+				family = v1.IPv6Protocol
```
`if netutils.IsIPv6(ip) {`

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -142,72 +142,97 @@ func Test_detectNodeIPs(t *testing.T) {
+			nodeIPOverride: []string{"10.0.0.1"},
+			ipFamilyPolicy: v1.IPFamilyPolicySingleStack,
+			expectedFamily: v1.IPv4Protocol,
+			expectedIPv4:   "10.0.0.1",
```
in cases where you're explicitly expecting a nil/empty result, I think it makes things clearer to say `expectedIPv6: ""` rather than leaving it unsaid.

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -142,72 +142,97 @@ func Test_detectNodeIPs(t *testing.T) {
+		{
+			name:           "No Valid IP found and no NodeIPOverride",
+			rawNodeIPs:     nil,
+			nodeIPOverride: []string{""},
```
`nodeIPOverride` should never be `[]string{""}`. If you don't specify it it would be `nil`.

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -142,72 +142,97 @@ func Test_detectNodeIPs(t *testing.T) {
 		},
 		{
-			name:           "No Valid IP found and unspecified bind address",
+			name:           "No Valid IP found and unspecified NodeIPOverride",
```
Maybe "and NodeIPOverride is unspecified address". Saying "unspecified NodeIPOverride" makes me think you mean `nil`, not `"0.0.0.0"`. (Yeah, it was sort of ambiguous before.)

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -217,8 +242,21 @@ func Test_detectNodeIPs(t *testing.T) {
+			expectedIPv6:   "2001:db8::2",
+		},
+		{
+			name: "Dual stack, primary IPv4 and RequiredDualStack IPFamilyPolicy",
```
typo "Require**d**DualStack"

(likewise below)

danwinship commented on cmd/kube-proxy/app/server_test.go:
```diff
@@ -272,23 +326,82 @@ func Test_detectNodeIPs(t *testing.T) {
+				netutils.ParseIPSloppy("90.90.90.90"),
+			},
+			nodeIPOverride: []string{"0.0.0.0", "::"},
+			expectedFamily: v1.IPv6Protocol,
```
Uh... this seems wrong. If the user says `nodeIPOverride: [ "0.0.0.0", "::" ]`, then they are saying they want IPv4-primary.

It seems this is consistent with the existing `--bind-address` behavior though. Ah, but that's probably because there's no such thing as `bindAddress: nil`; it defaults to `"0.0.0.0"`, so we can't treat `"0.0.0.0"` as being an actual opinion from the user.

Ugh... ok, I want to think about this again. I think we need to just have different, incompatible node-IP-detecting behavior for v1alpha1 and v1alpha2 configs.

danwinship replied:
Sorry, I never got back to this. So, I think the fix for now is to just have both `bindAddress` and `nodeIPOverride` in the internal version of the config, where only `bindAddress` would get filled in for CLI args or v1alpha1 config, and only `nodeIPOverride` would get filled in for v1alpha2 config. Then the kube-proxy startup code can use the backward-compatible behavior for CLI/v1alpha1 configs, and the better, more-explicit behavior for v1alpha2 configs
